### ISSUE TITLE
Joystick ids changing

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#494]
When adding or removing physical joysticks, the ids change and therefore the mappings too.

![image](https://user-images.githubusercontent.com/24360774/63003739-2609af80-be79-11e9-8f41-38d6c5eea756.png)

Adding Stelladaptors also changes the IDs of already added joysticks (TODO: screenshot).

If we cannot rely on IDs returned by SDL here, maybe we should work with hashes (controller name and properties) instead.

**Comment by sa666666** [comment]
This is directly from the SDL2 docs.  The joystick ID is a strictly increasing integer, and is incremented each time a joystick is added to the system, but never decremented.  So the behaviour you're seeing is exactly what should be happening.

We cannot rely on ID at all.  It is not supported across multiple runs of Stella, nor even during the same run, as an unplug-replug cycle will change it.  This is why in the original joystick mapping code I saved mappings with the _joystick USB name_ returned by Stella, **not** by integer ID.  Note that we _can_ use an ID, we just can't save it to the settings, since it is meaningless on the next run.

Each time a stick is inserted, the internal stick database needs to be queried, and the mappings set up for that stick with whatever ID it happens to use at that time.  If it's later unplugged, then the mappings are still saved with the stick, but by name, not integer ID.  Later, if it's inserted again, the same thing needs to happen with a new ID.

This is not as hard as it might appear.  We just need to make sure the joystick _names_ are unique, and I've already added code to do that.

**Comment by thrust26** [comment]
Maybe the controller name is not good enough. I can't test at the moment, but I suppose a Stelladaptor has always the same name, but it can support joysticks, paddles, trackballs etc.  Also I am not sure if the names are unique (see above).

That's why I suggested using hashs above. That should make things more unique.

**Comment by sa666666** [comment]
Stelladaptor / 2600-daptor definitely have the same name.  It doesn't expose (through SDL at least) what the actual controller is that's currently plugged in.

I completely forgot about there now essentially being _groups_ of mappings for each joystick.  So I guess you have to hash names.  I will leave this to you for now, as you wrote most of that code ...

**Comment by thrust26** [comment]
I thought that ID part is from you (or someone else not me).

**Comment by sa666666** [comment]
No, I meant that using hashing is something you introduced.  TBH, you've added so much stuff there that you've basically wrote more of it than me at this point.

**Comment by sa666666** [comment]
Hmm, now I'm confused.  Are the ID's changing affecting any of the mapping work you did, or is it just cosmetic?  I would need to see what bugs are coming from it before I can attempt to fix it.

**Comment by thrust26** [comment]
They affect the mapping. 

Example 1:
- start Stella
- plugin a joystick 0
- start game
- map button 1 to action A (displayed as J0/B1)
- test (it works)
- all other mappings say "J0/.."
- unplug the joystick and plugin again
- test 1st mapping and other old joystick mappings (do **not** work anymore)
- check Input dialog: initially the mapping is still displayed as "J**0**/B1", but when you change selection this gets updated to "J1/B1", all other mappings show J1 too.
- map button 2 to action B (displayed as J1/B2)
- test 2nd mapping (works)
- test 1st mapping (does not work, and all other joystick still mappings too)
- note that all other mappings now say "J1/.."
- exit to launcher and start game again
- now **all** mappings start working again
- also after restarting Stella

So something in the hot plugging detection seems broken.  I will check if the error is on my side and report back.

**Comment by thrust26** [comment]
@sa666666 Since I am currently on vacation I didn't take all my equipment with me (but my **office** notebook and a joystick :smile:), so I cannot test Stelladaptor behavior. IIRC there were ID problems too and probably also the potential problems with different controller emulation. Can you test that for me?

**Comment by sa666666** [comment]
I am waiting on another 2600-daptor to arrive, since a few people have reported problems with having two plugged in at the same time.  I will test that stuff together, when I have both daptor's available.

**Comment by thrust26** [comment]
Cool. Which ...daptor models will you own then?

**Comment by sa666666** [comment]
I have the original Stelladaptor.  And each of the 2600-daptor ones (I, II, and D9).  I have a second D9 coming.

**Comment by thrust26** [comment]
@sa666666 I tested today, and there seems to be no major issue with changing IDs anymore. 

The only problem I see is, that currently ID % 1 defines the joystick port. IMO we should allow the user to assign a physical controller to an emulated port.

**Comment by sa666666** [comment]
@thrust26, is this any longer an issue, or can we close it?

**Comment by thrust26** [comment]
No issue, but a missing feature as described in my previous post. Should we create a new issue for this enhancement?

**Comment by sa666666** [comment]
Yeah, maybe another issue that is specific to what we're looking for here.

**Comment by thrust26** [comment]
Created dedicated issue for port selection enhancement #897.
